core/types.ml

```diff
@@ -3086,6 +3791,29 @@ module RoundtripPrinter : PRETTY_PRINTER = struct
             | None | Some Absent | Some (Meta _) -> false
             | _ -> raise tag_expectation_mismatch
           in
+          let decide_skip ctx vid =
+            let anonymity = get_var_anonymity ctx vid in
+            if Context.implicit_shared_effect_exists ctx
+            then begin
+                let es_policy = Policy.es_policy (Context.policy ctx) in
+                let module ES = Policy.EffectSugar in
+
+                let arrows_show_impl_shared = ES.arrows_show_the_one es_policy in
+                let arrows_collection_fresh = ES.arrows_collection_fresh es_policy in
+
+                match anonymity with
+                | Visible -> false (* decided Visible, cannot skip *)
+                | Anonymous -> arrows_show_impl_shared ||
+                                 (arrows_collection_fresh && Context.is_ambient_arrow_curried ctx)
+                | ImplicitEff -> not arrows_show_impl_shared
+              end
+            else match anonymity with
+                 | Anonymous -> true (* skip *)
+                 | Visible   -> false (* no skip *)
+                 | _         ->
+                    (* this is not supposed to happen if there is no shared effect variable *)
+                    raise tag_expectation_mismatch
```

Comment:
raise `internal_error` here. In general `tag_expectation_mismatch` should only be used for things that match over the structure of types.